### PR TITLE
SUP-2585: Add Support for changing Build Drivers

### PR DESCRIPTION
**Pull request by tomowatt**
Per Issue: https://github.com/buildkite-plugins/docker-compose-buildkite-plugin/issues/457

To be able to fully support using the `cache-to` parameter, the Build Driver needs to be changed to use a different driver other than the default `docker` driver.
This PR will add the basic ability to change the driver used in a per Job basis and clean up the Build Instance afterwards to avoid issues of lingering Build Instances.

- [x] README.md updated
- [x] tests added
- [x] examples added to examples/examples.md

### CODE REVIEW

**Comment by tomowatt** [comment]
Bringing out of Draft for initial reviews whilst adding tests

File: hooks/pre-command [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+builder_use="$(plugin_read_config BUILDER_USE "false")"
+builder_name="$(plugin_read_config BUILDER_NAME "")"
+
+if [[ "${builder_create}" == "true" ]] || [[ "${builder_use}" == "true" ]]; then
+    if [[ -z "${builder_name}" ]]; then
+        echo "+++ 🚨 Builder Name cannot be empty when using 'create' or 'use' parameters"
+        exit 1
+    fi
+    # exporting to use in pre-exit
+    export builder_name
```
Comment (toote) [inline hooks/pre-command:19]: why is this necessary? couldn't you just use `$(plugin_read_config BUILDER_NAME "")` in the pre-exit hook as well?

File: README.md [hunk 2]
```diff
@@ -344,6 +344,76 @@ If set to true, docker compose will build with the `--with-dependencies` option
...
+
+##### `driver-opt` (string)
+
+Optional, commas separated, Key-Value pairs of driver-specific options to configure the Builder Instance when using `create`. Available options for each Driver:
+
+- [docker-container](https://docs.docker.com/build/builders/drivers/docker-container/)
+- [kubernetes](https://docs.docker.com/build/builders/drivers/kubernetes/)
+- [remote](https://docs.docker.com/build/builders/drivers/remote/)
+
+Example: `memory=100m`
```
Comment (toote) [inline README.md:387]: I would make the example use more than one option to show how to comma-separate them

File: plugin.yml [hunk 3]
```diff
@@ -23,6 +23,30 @@ configuration:
...
+        name:
+          type: string
+        platform:
+          type: string
+        remote-address:
+          type: string
+        remove:
+          type: boolean
+        use:
+          type: string
```
Comment (toote) [inline plugin.yml:49]: I would suggest adding `additionalProperties: false` to restrict valid elements to exactly those listed

**Review by toote — APPROVED** [review]
Despite all my comments, this is a very thorough implementation of the functionality :tada:

**Comment by tomowatt** [comment]
thanks @toote will do the clean up and of the docs and a bit of refactoring/changing where it makes sense, went down the DRY route with some ideas but definitely make sense to make it more clear in some situations with the `pre-exit` hook

File: hooks/pre-command (hunk 1 above)
Comment (tomowatt) [reply hooks/pre-command:19]: trying to be DRY but yeah makes it easier and clearer to have the `pre-exit` hook do environent variable reading

File: hooks/pre-command [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+        build_driver="$(plugin_read_config BUILDER_DRIVER "")"
+        valid_drivers="docker-container|kubernetes|remote"
+        if [[ "${build_driver}" =~ ^(${valid_drivers})$ ]]; then
+            builder_instance_args=(--driver "${build_driver}")
+        else
+            echo "+++ 🚨 Invalid driver: '${build_driver}'"
+            echo "Valid Drivers: ${valid_drivers//|/, }"
+            exit 1
+        fi
```
[the remark this answers is not in the document]
Comment (tomowatt) [reply hooks/pre-command:34]: Yeah feeling the same too as well, haven't came across other Driver documentation that's wildly available though aware there is the `cloud` 

```
$ docker buildx create --help
...
...
    --driver string             Driver to use (available: "cloud", "docker-container", "kubernetes", "remote")
...
...
```

File: docs/examples.md [hunk 5]
```diff
@@ -312,3 +312,111 @@ steps:
...
+      - docker-compose#v5.4.1:
+          build: app
+          push: app:index.docker.io/myorg/myrepo:my-branch
+          cache-from:
+            - "app:myregistry:port/myrepo/myapp:my-branch"
+            - "app:myregistry:port/myrepo/myapp:latest"
+          builder:
+            name: container
+            remove: true
+```
```
Comment (toote) [inline docs/examples.md:378]: Based on the previous comment, this example would not actually create or use the builder defined :S

File: docs/examples.md [hunk 6]
```diff
@@ -312,3 +312,111 @@ steps:
...
+
+**Removing a Builder Instance by default will remove the daemon running it and its state (which can be used for cache).**
+**To keep the daemon or state, use the `keep-daemon` or `keep-state` parameters.**
+**These parameter are only applicable with specific Drivers, for detail see [`docker buildx rm`](https://docs.docker.com/reference/cli/docker/buildx/rm/).**
+
+### Reusing caches from remote regestries
+
+A newly spawned agent won't contain any of the docker caches for the first run which will result in a long build step. To mitigate this you can reuse caches from a remote registry, but requires pushing cache and manifests to a registry using a Builder Driver that supports cache exports e.g., `docker-container` - the `docker` driver does not support this by default. This requires use of the `cache-from`, `cache-to`, `name` and `use` parameters but will use the `create` and `driver` parameters to create the Builder Instance across multiple Agents:
```
Comment (toote) [inline docs/examples.md:386]: you may want to add a clarification here about remote registries that require authentication

**Review by toote — APPROVED** [review]
:clap: very thorough second round... most of my comments are cosmetic or enhancements to what is already there so feel free to leave them for future improvements

File: tests/builder-instances.bats [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+load "${BATS_PLUGIN_PATH}/load.bash"
+load '../lib/shared'
+
+# export DOCKER_STUB_DEBUG=/dev/tty
+
+teardown() {
+    if [[ -f "${BATS_MOCK_BINDIR}/docker" ]]; then
+        unstub docker
+    fi
+}
```
[the remark this answers is not in the document]
Comment (tomowatt) [reply tests/builder-instances.bats:12]: originally used it as a quick win for running tests locally, but was testing only the new test files and all worked as should.
Re-testing with all the test files locally, getting errors with unstubbing, so will clean up. Interestingly that doesn't occur when the tests are run on an Agent.

File: hooks/pre-command [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+if [[ "${builder_create}" == "true" ]]; then
+    if ! builder_instance_exists "${builder_name}"; then
+        builder_instance_args=()
+
+        build_driver="$(plugin_read_config BUILDER_DRIVER "")"
+        valid_drivers="docker-container|kubernetes|remote"
+        if [[ "${build_driver}" =~ ^(${valid_drivers})$ ]]; then
+            builder_instance_args=(--driver "${build_driver}")
+        else
```
[the remark this answers is not in the document]
Comment (tomowatt) [reply hooks/pre-command]: will leave the current logic for now so it can be revisited in future if required as other drivers may also require additional parameters as well
